---
title: "Disease GWAS colocalisation analysis"
output: html_notebook
---

We wish to link our immune cell csQTLs to human diseases. To that end I'll perform a colocalisation analysis between GWAS catalogue autoimmune and infectious disease trait results (including 
COVID19).

```{r}
suppressMessages({
library(SingleCellExperiment)
library(miloR)
library(Matrix)
library(MatrixGenerics)
library(ggplot2)
library(ggsci)
library(ggrepel)
library(cowplot)
library(scales)
library(scattermore)
library(BiocParallel)
library(MatrixEQTL)
library(biomaRt)
library(stringr)
library(dplyr)
library(coloc)
library(VennDiagram)
library(scico)
library(viridis)
library(ggtext)
library(glue)
library(coloc)
library(readr)
})
```

# Read in Milo object

```{r}
rand.ni.milo <- readRDS("~/Dropbox/GLMM/data/Randolph_Milo.RDS")
rand.ni.meta <- as.data.frame(colData(rand.ni.milo))
rand.ni.meta$CellID <- colnames(rand.ni.milo)
rand.ni.fr <- as.data.frame(reducedDim(rand.ni.milo, "UMAP"))
rand.ni.fr$CellID <- colnames(rand.ni.milo)

rand.ni.meta <- merge(rand.ni.meta, rand.ni.fr, by="CellID")
```

```{r}
# use this for annotating the csQTL results
glmm.da.res <- read.table("~/Dropbox/GLMM/results/Randolph_MockGLMM_results.tsv",
                      sep="\t", header=TRUE, stringsAsFactors = FALSE)
```

# Extract lead SNPs

```{r}
# gwas results files - do these only contain the nhoods for which the SNP is DA?
gwas.dir <- "~/Dropbox/GLMM/locus.dir/"
gwas.files <- list.files(gwas.dir, pattern="\\.gz$", full.names=TRUE, recursive=TRUE, include.dirs=TRUE)
gwas.list <- list()
for(x in seq_along(gwas.files)){
  x.file <- gwas.files[x]
  x.lead <- gsub(x.file, pattern="([[:ascii:]]*)/(chr[0-9]+)/([0-9]+)_([0-9]+)_([ATCG]+)_([ATCG]+)(_locus\\.tsv\\.gz)", 
                 replacement="\\3_\\4_\\5_\\6", perl=TRUE)
  x.chr <- gsub(x.lead, pattern="([0-9]+)_([0-9]+)_([ATCG]+)_([ATCG]+)", replacement="\\1")
  x.bp <- as.numeric(gsub(x.lead, pattern="([0-9]+)_([0-9]+)_([ATCG]+)_([ATCG]+)", replacement="\\2"))
  
  # tabix this for speed
  x.tab.query <- paste0("tabix ", x.file, " ", paste0(x.chr, ":", x.bp-10, "-", x.bp+10))
  x.df <- do.call(rbind.data.frame, lapply(system(x.tab.query, intern=TRUE), FUN=function(TX){
    unlist(strsplit(TX, split="\t", fixed=TRUE))
  }))
  
  colnames(x.df) <- c("SNP", "BP", "CHR", "logFC", "SE", "GeneticVariance", "SpatialFDR", "Nhood", "ALT", "REF")
  x.df$CHR <- as.numeric(x.df$CHR)
  x.df$BP <- as.numeric(x.df$BP)
  x.df$logFC <- as.numeric(x.df$logFC)
  x.df$SE <- as.numeric(x.df$SE)
  x.df$GeneticVariance <- as.numeric(x.df$GeneticVariance)
  x.df$SpatialFDR <- as.numeric(x.df$SpatialFDR)
  x.df$Nhood <- as.numeric(x.df$Nhood)
  
  # only keep the FDR <= 1e-8, there are too many SNPs otherwise
  x.df <- x.df[x.df$SpatialFDR <= 1e-8, ]
  x.df <- merge(x.df, glmm.da.res[, c("Nhood", "ident", "ident_fraction", "NhoodGroup")])
  x.df$LeadSNP <- x.lead
  # only keep the lead SNP
  x.df <- x.df[x.df$SNP == x.df$LeadSNP, , drop=FALSE]
  gwas.list[[x.file]] <- x.df
}

gwas.hits.df <- do.call(rbind.data.frame, gwas.list)
lead.snps <- unique(gwas.hits.df$LeadSNP)
table(gwas.hits.df$ident) 
```

# GWAS summary statistic colocalisation

We have summary statistics from the GWAS catalogue and from the COVID19 HGI (release 7): https://www.covid19hg.org/. Both sets of summary statistics are tabix-indexed for efficient  
access. In general, the GWAS catalogue only has 1 or 2 variants for each locus, so these will need to be analysed differently.

__NB__: As this is slow I ran it on the HPC. The code is here for reference.

```{r, eval=FALSE}
# gwas.cat.file <- "~/Dropbox/GLMM/COVID19_HGI/gwas_catalog.tsv.bgz"
covid.hgi.file <- "~/Dropbox/GLMM/COVID19_HGI/COVID19_HGI_A2_ALL_leave_23andme_20220403.tsv.gz"

# gwas.col.names <- colnames(readr::read_tsv(gwas.cat.file, n_max = 1, show_col_types=FALSE)) 
covid.col.names <- colnames(readr::read_tsv(covid.hgi.file, n_max = 1, show_col_types=FALSE)) 

# this code is here for reference - this analysis was run on the HPC
# gwas.csqtl.coloc.list <- list()
covid.csqtl.coloc.list <- list()

interval.size <- 500000L # 1Mb interval around the lead SNP
# compute the nhood standard deviations
rand.ni.sd <- apply(nhoodCounts(rand.ni.milo), 1, sd) # these are numbered sequentially

# lead.snps are the ones that are _not_ in gene deserts
for(x in seq_along(lead.snps)){
  x.snp <- lead.snps[x]
  x.nhoods <- unique(gwas.hits.df[gwas.hits.df$LeadSNP %in% x.snp, ]$Nhood)
  
  x.chr <- gsub(x.snp, pattern="([0-9]+)_([0-9]+)_([ATCG]+)_([ATCG]+)", 
                replacement="\\1")
  x.bp <- as.numeric(gsub(x.snp, pattern="([0-9]+)_([0-9]+)_([ATCG]+)_([ATCG]+)", 
                          replacement="\\2"))
  
  x.csqtl.file <- gwas.files[grepl(gwas.files, pattern=x.snp)] # these are from the first code block in this notebook
  if((x.bp - interval.size) > 1){
    x.tabix.query <- paste0("tabix ", x.csqtl.file, " ", x.chr, ":", x.bp-interval.size, "-", x.bp+interval.size) # interval size is also from above
  } else{
    x.tabix.query <- paste0("tabix ", x.csqtl.file, " ", x.chr, ":", 1, "-", x.bp+interval.size) # interval size is also from above
  }
  x.csqtl.df <- do.call(rbind.data.frame, lapply(system(x.tabix.query, intern=TRUE), FUN=function(TX){
    unlist(strsplit(TX, split="\t", fixed=TRUE))
    }))

  colnames(x.csqtl.df) <- c("SNP", "BP", "CHR", "logFC", "SE", "GeneticVariance", "SpatialFDR", "Nhood", "ALT", "REF")
  x.csqtl.df$CHR <- as.numeric(x.csqtl.df$CHR)
  x.csqtl.df$BP <- as.numeric(x.csqtl.df$BP)
  x.csqtl.df$logFC <- as.numeric(x.csqtl.df$logFC)
  x.csqtl.df$SE <- as.numeric(x.csqtl.df$SE)
  x.csqtl.df$GeneticVariance <- as.numeric(x.csqtl.df$GeneticVariance)
  x.csqtl.df$SpatialFDR <- as.numeric(x.csqtl.df$SpatialFDR)
  x.csqtl.df$Nhood <- as.numeric(x.csqtl.df$Nhood)
  
  # HGI COVID19 results
  ## tabix query covid19 hgi results
  ## GWAS catalogue generally only has a single variant reported - not enough for coloc.
  x.covid.tabix <- paste0("tabix ", covid.hgi.file, " ", x.chr, ":", x.bp-interval.size, "-", x.bp+interval.size) # interval size is also from above
  x.covid.df <- do.call(rbind.data.frame, lapply(system(x.covid.tabix, intern=TRUE), FUN=function(TX){
    unlist(strsplit(TX, split="\t", fixed=TRUE))
    }))
  colnames(x.covid.df) <- covid.col.names
  x.covid.df$SNP <- gsub(x.covid.df$SNP, pattern=":", replacement="_")
  x.covid.df$all_inv_var_meta_beta <- as.numeric(x.covid.df$all_inv_var_meta_beta)
  x.covid.df$all_inv_var_meta_sebeta <- as.numeric(x.covid.df$all_inv_var_meta_sebeta)
  x.covid.df$all_inv_var_meta_cases <- as.numeric(x.covid.df$all_inv_var_meta_cases)
  x.covid.df$all_inv_var_meta_controls <- as.numeric(x.covid.df$all_inv_var_meta_controls)
  x.covid.df$N <- x.covid.df$all_inv_var_meta_case + x.covid.df$all_inv_var_meta_controls
  
  x.covid.df <- x.covid.df[!is.na(x.covid.df$all_inv_var_meta_beta), ]
  x.covid.df <- x.covid.df[!is.na(x.covid.df$all_inv_var_meta_sebeta), ]
  x.covid.df <- x.covid.df[!is.na(x.covid.df$N), ]
  x.covid.df <- x.covid.df[!duplicated(x.covid.df$SNP), ]
  
  for(q in seq_along(x.nhoods)){
      q.nh <- x.nhoods[q]
      q.csqtl <- x.csqtl.df[x.csqtl.df$Nhood %in% q.nh, ]
      q.csqtl <- q.csqtl[!is.na(q.csqtl$logFC), ]
      
      x.coloc.list <- list("beta"=q.csqtl$logFC, "varbeta"=(q.csqtl$SE)**2,
                           "snp"=q.csqtl$SNP, "position"=q.csqtl$BP,
                           "type"="quant", "sdY"=rand.ni.sd[q.nh])
      
      x.cov.overlap <- intersect(x.covid.df$SNP, x.csqtl.df$SNP)
      if(length(x.cov.overlap) > 5){
          x.covid.coloc.list <- list("beta"=x.covid.df$all_inv_var_meta_beta,
                                     "varbeta"=(x.covid.df$all_inv_var_meta_sebeta)**2,
                                     "snp"=x.covid.df$SNP,
                                     "position"=x.covid.df$POS,
                                     "type"="cc")
          # check coloc validity
          if(is.null(check_dataset(x.covid.coloc.list)) & is.null(check_dataset(x.coloc.list))){
            k.coloc <- coloc.abf(x.coloc.list, x.covid.coloc.list)
            k.df <- do.call(cbind.data.frame, as.list(k.coloc$summary))
            k.df$Nhood <- q.nh
            k.df$PhenoDescription <- "COVID19.HGI"
            k.df$LeadSNP <- x.snp

            k.ppsum <- sum(k.df[c(2:6)])
            if(k.ppsum > 1){
              warning("PP sum > 1 for COVID19 in nhood ", q.nh, " for SNP ", x.snp)
            }

            covid.csqtl.coloc.list[[paste0(x.snp, "_", q.nh, "_", "COVID19")]] <- k.df

          }
      }
  }
}



csqtl.covid.coloc.df <- do.call(rbind.data.frame, covid.csqtl.coloc.list)
# merge with Randolph cell type annotations
csqtl.covid.coloc.df <- merge(csqtl.covid.coloc.df, glmm.da.res[, c("Nhood", "NhoodGroup", "ident", "ident_fraction")], by="Nhood", all.x=TRUE)

write.table(csqtl.covid.coloc.df,
            file="~/Dropbox/GLMM/results/COVID19_csQTL_coloc.tsv",
            quote=FALSE, sep="\t", row.names=FALSE)
```



